src/database/RecipeEntity.js

```diff
@@ -1,3 +1,84 @@
 const {Datastore} = require("@google-cloud/datastore");
 const datastore = new Datastore();
 
+async function getRecipeEntities(pageCursor) {
+  let query = datastore.createQuery("Recipe");
+
+  if (pageCursor) {
+    query = query.start(pageCursor);
+  }
+
+  const results = await datastore.runQuery(query);
+  const entities = results[0];
+  const info = results[1];
+
+  if (info.moreResults !== Datastore.NO_MORE_RESULTS) {
+    const results = await getRecipes(info.endCursor);
+    return [...entities, results[0]];
+  }
+  return entities;
+}
+
+async function getRecipes(data) {
+  if (!data.recipes) {
+    const entities = await getRecipeEntities();
+    const obj = entities.map(recipe => {
+      return {
+        id: recipe[datastore.KEY].id,
+        name: recipe.name,
+        ingredients: recipe.ingredients,
+        method: recipe.method,
+        createdBy: recipe.createdBy,
+        time: recipe.time
+      };
+    });
+    data.recipes = JSON.stringify(obj);
+  }
+  let recipes = JSON.parse(data.recipes);
+  return filterRecipes(data, recipes);
```

Comment:
This implies that you are getting ALL the recipes from your database and then doing filtering one by one through code. This is incredibly inefficient and taxing on memory (what if you've 4 billion recipes? How to keep all of them in memory when array index is going to go out of bounds?)

Instead, you should make sure the results you get from the database is already filtered. Your database query should support things like limit, filter, groupby, start, etc. You can also select specific columns rather than get everything.

e.g. datastore.createQuery("Recipes").limit(10).filter('name', '=', 'Eggfooyoung')

Look into sql queries and capabilities if this doesn't make sense to you. The nodejs api reference for query is at https://cloud.google.com/nodejs/docs/reference/datastore/latest/datastore/query#_google_cloud_datastore_Query_filter_member_1_